gh-141070: Add PyUnstable_Object_Dump() function (#141072)
python/cpython: Doc/c-api/object.rst

   instead of the :func:`repr`.

> .. c:function:: void PyObject_Dump(PyObject* op)

Reviewer: ```suggestion
.. c:function:: void PyObject_Dump(PyObject *op)
```

>    Dump an object *op* to ``stderr``. Function used for debugging.

Reviewer 2: ```suggestion
   Dump an object *op* to ``stderr``. This should only be used for debugging.
```

   It can be called without an :term:`attached thread state`, even if it's not

>    recommended.

Reviewer 2: ```suggestion
   This function can be called without an :term:`attached thread state`, but it's not
   recommended to do so.
```

>    Implement an heuristic to detect if the object memory has been freed.

Reviewer 2: ```suggestion
   Implement a heuristic to detect if the object memory has been freed.
```

>    Example of output:

Reviewer 3: ```suggestion
   The output is intended to try dumping objects even after memory corruption:

   - Information is written starting with fields that are the least likely to crash when accessed
   - Less information is printed if Python detects that the object has been freed
   - This function can be called without an :term:`attached thread state`, but this may cause race conditions, deadlocks or crashes
   - An object that does not belong to the current interpreter may be dumped, but this may also cause crashes or unintended behavior
   
   The output format may change at any time.

   Example of output:
```
